Public interfaces for dashboard app

ArtificialOwl opened the pull request.

rullzer commented on lib/public/Dashboard/IDashboardWidget.php:
```diff
@@ -0,0 +1,53 @@
+<?php
+
```
Licence header

rullzer commented on lib/public/Dashboard/IDashboardWidget.php:
```diff
@@ -0,0 +1,53 @@
+	/**
+	 * @return string
+	 */
+	public function getId();
```
NC14 is >php 7.0. So you can use return types

daita replied:
can I use : void ? (php 7.1)

rullzer commented on lib/public/Dashboard/IDashboardWidget.php:
```diff
@@ -0,0 +1,53 @@
+<?php
```
strict typing since we are >7.0 anyway

daita commented:
@rullzer: please have a look to the asked edit

nickvergessen commented on lib/public/Dashboard/Model/IWidgetEvent.php:
```diff
@@ -0,0 +1,130 @@
+	 *
+	 * @param string $widgetId
+	 */
+	public function __construct(string $widgetId);
```
I wouldn't add the constructor to the interfaces.
Just call `setWidgetId` instead

juliushaertl commented:
@daita You will also need to adjust the following xsd files to make the dashboard section valid in the appinfo.xml:

https://github.com/nextcloud/server/blob/master/resources/app-info-shipped.xsd
https://github.com/nextcloud/server/blob/master/resources/app-info.xsd

And the app store needs those updates as well.

juliushaertl commented:
Just a general thing, when we add new sections to the appinfo.xml like <dashboard> for this case, does this break apps that also support versions previous to 14? 

The app store only has one version-independent file to validate the appinfo, so that should work fine, but does the server also checks for a valid appinfo at some point?

nickvergessen commented:
Yes the app check-code checks it, but additional stuff is not a problem as far as I know.
Only if you have stuff, which is not parsing valid, it will cry

MorrisJobke commented on lib/public/Dashboard/IDashboardWidget.php:
```diff
@@ -0,0 +1,79 @@
+interface IDashboardWidget {
+
+	/**
+	 * @return string
```
Missing `@since 14.0.0` on the methods and the interface.

MorrisJobke commented on lib/public/Dashboard/IDashboardWidget.php:
```diff
@@ -0,0 +1,79 @@
+<?php declare(strict_types=1);
```
Put the `declare()` on a new line.

MorrisJobke commented:
It's a bit late for that ... freeze was on Friday ... at least for the server PR.

And also it was never brought up to be needed for 14. :/

jancborchardt commented:
No need to rush in the dashboard, no? We can merge it very early for 15, and then properly work on adding endpoints to the different apps and have the Dashboard app use those, like any other app.

daita commented:
Well, you know. After working on something else for months and finally having few days to do something fun. But I get it; @nickvergessen already told me last week that I should not waste my time on it.

So, no rush.

daita commented:
So, ready for NC15.

@nickvergessen : Added also a IDashboardManager that can be called with Automatic DI so there is no more API, just call the right method from the dashboardManager, that will check if app is installed or not.

nickvergessen commented on lib/private/Dashboard/DashboardManager.php:
```diff
@@ -0,0 +1,134 @@
+	 * @throws DashboardAppNotAvailableException
+	 */
+	public function createGlobalEvent(string $widgetId, array $payload, string $uniqueId = ''
+	) {
```
opsy

nickvergessen commented on resources/app-info-shipped.xsd:
```diff
@@ -654,6 +656,14 @@
+                        maxOccurs="unbounded"/>
+            maxOccurs="unbounded"/>
+        </xs:sequence>
+    </xs:complexType>
```
this needs to go to the appstore too

rullzer replied:
More general why is this needed here? I don't see it being used yet.

ChristophWurst commented on lib/public/Dashboard/IDashboardWidget.php:
```diff
@@ -0,0 +1,101 @@
+	/**
+	 * @since 15.0.0
+	 *
+	 * @return array
```
Please add more documentation to these methods. What are they supposed to return? `array` is very generic :wink:

ChristophWurst commented on lib/public/Dashboard/IDashboardWidget.php:
```diff
@@ -0,0 +1,101 @@
+	/**
+	 * @since 15.0.0
+	 *
+	 * @return array
```
same as above

ChristophWurst commented on lib/public/Dashboard/Model/IWidgetRequest.php:
```diff
@@ -0,0 +1,128 @@
+	 *
+	 * @param string $widgetId
+	 */
+	public function __construct(string $widgetId);
```
the interface determines the constructor interface? This doesn't seem right :see_no_evil:

ChristophWurst commented on lib/public/Dashboard/Model/IWidgetRequest.php:
```diff
@@ -0,0 +1,128 @@
+	/**
+	 * @since 15.0.0
+	 *
+	 * @param string $request
```
what kind of request is this string?

MorrisJobke replied:
@daita 🏓

ChristophWurst replied:
:man_shrugging: :ping_pong:

ChristophWurst commented on lib/public/Dashboard/Model/IWidgetSettings.php:
```diff
@@ -0,0 +1,135 @@
+	 * @param string $userId
+	 * @param string $widgetId
+	 */
+	public function __construct(string $widgetId, string $userId);
```
Please remove this. Same issue as above.

MorrisJobke replied:
@daita 🏓

daita replied:
saw that ... removed.

ChristophWurst reviewed (dismissed):
Some nitpicks. Please also make sure you have a line break at the end of each file (as github shows here) to be consistent with our other source files.

Thanks!

rullzer commented on resources/app-info.xsd:
```diff
@@ -642,6 +644,14 @@
+        <xs:sequence>
+            <xs:element name="widget" type="php-class" minOccurs="1"
+                        maxOccurs="unbounded"/>
+                        maxOccurs="unbounded"/>
```
that doesn't look right.

rullzer commented on resources/app-info-shipped.xsd:
```diff
@@ -654,6 +656,14 @@
+        <xs:sequence>
+            <xs:element name="widget" type="php-class" minOccurs="1"
+                        maxOccurs="unbounded"/>
+            maxOccurs="unbounded"/>
```
same here

nickvergessen commented on lib/private/Dashboard/DashboardManager.php:
```diff
@@ -0,0 +1,134 @@
+		}
+
+		try {
+			return \OC::$server->query($service);
```
you can inject `IServerContainer` in your constructor. Then you can also better unit test this.

daita commented:
:ping_pong: 

Would be great to go on with this one, as it is a blocker to work on the dev of widgets for 15.
(I know it's missing tests, this would come later)

MorrisJobke commented on lib/public/Dashboard/IDashboardWidget.php:
```diff
@@ -0,0 +1,163 @@
+	 *   'content'  => (string) path to the HTML Template of the widget,
+	 *   'function' => (string) JavaScript function to be called when
+	 *                          loading the widget on the dashboard
+	 * ];
```
A template object would be nice, because it is automatically properly typed and has then proper autocompletion as well. 😉

daita replied:
moved to objects in various spot, pushed. Will comment during the day.

MorrisJobke commented on lib/public/Dashboard/IDashboardWidget.php:
```diff
@@ -0,0 +1,163 @@
+	 * 	  ]
+	 *
+	 * 'push' contains the JS function to be called on push event:
+	 * 	  'push' => (string) JavaScript function to be called on push event
```
A settings object would be nice instead of a multidimensional array. 🤔

MorrisJobke reviewed (commented):
Beside the comments it looks good 👍

ChristophWurst commented:
Could you briefly describe how the new interfaces fit together please? I still haven't quite understood how this is supposed to work. Would it make sense to allow more than one entry in `info.xml` per app?

ChristophWurst commented:
- [ ] Since the info schema here is a copy of the appstore one, we should get https://github.com/nextcloud/appstore/pull/567 in first.

ChristophWurst commented:
> Could you briefly describe how the new interfaces fit together please?

Sorry for being so nitpicky here, but since this will be a quite large public API of our server which we hope lots of apps use, I really want this to be as easy to use as possible (also in regards to documentation).

daita commented:
All commented and looking good. 

Please review !

ChristophWurst commented on lib/private/Dashboard/DashboardManager.php:
```diff
@@ -0,0 +1,144 @@
+	 */
+	public function getWidgetConfig(string $widgetId, string $userId): IWidgetConfig {
+		/** @var \OCA\Dashboard\Service\WidgetsService $widgetsService */
+		$widgetsService = $this->getService('\OCA\Dashboard\Service\WidgetsService');
```
please use the `::class` syntax, this makes finding class usages and refactoring later a lot easier :+1:

nickvergessen replied:
Would require another try-catch in here of course, in case the dashboard app is not here.

ChristophWurst replied:
Why would it? `::class` is simply replaced by the FQCN (string) by the interpreter.

nickvergessen replied:
I thought this would invoke autoloading and therefor throw? 🤷

daita replied:
The thing is that I am not sure the app is available at this point, and calling the ::class does not clean enough

ChristophWurst replied:
```
$ php -a
Interactive shell

php > echo Joas::class;
Joas
```

ChristophWurst commented on lib/private/Dashboard/DashboardManager.php:
```diff
@@ -0,0 +1,144 @@
+	 */
+	public function createUsersEvent(string $widgetId, array $users, array $payload, string $uniqueId = '') {
+		/** @var \OCA\Dashboard\Service\EventsService $eventsService */
+		$eventsService = $this->getService('\OCA\Dashboard\Service\EventsService');
```
same

ChristophWurst commented on lib/private/Dashboard/DashboardManager.php:
```diff
@@ -0,0 +1,144 @@
+	 */
+	public function createGroupsEvent(string $widgetId, array $groups, array $payload, string $uniqueId = '') {
+		/** @var \OCA\Dashboard\Service\EventsService $eventsService */
+		$eventsService = $this->getService('\OCA\Dashboard\Service\EventsService');
```
same

ChristophWurst commented on lib/private/Dashboard/DashboardManager.php:
```diff
@@ -0,0 +1,144 @@
+	 */
+	public function createGlobalEvent(string $widgetId, array $payload, string $uniqueId = '') {
+		/** @var \OCA\Dashboard\Service\EventsService $eventsService */
+		$eventsService = $this->getService('\OCA\Dashboard\Service\EventsService');
```
same

ChristophWurst commented on lib/public/Dashboard/Model/IWidgetTemplate.php:
```diff
@@ -0,0 +1,232 @@
+	public function getInitFunction(): string;
+
+	/**
+	 * JavaScript function to be called when loading the widget on the
```
how will that function be invoked? Does it have to exist globally?

daita replied:
it does not have to exist globally, it is called by the app

ChristophWurst commented on lib/private/Dashboard/Model/WidgetTemplate.php:
```diff
@@ -0,0 +1,266 @@
+	 *
+	 * @return array
+	 */
+	public function getCss(): array {
```
how will this CSS be loaded? Is this property the actual CSS contents or just the URL?

rullzer commented on lib/private/Dashboard/Model/WidgetSetting.php:
```diff
@@ -0,0 +1,203 @@
+	 *
+	 * @param string $type
+	 */
+	public function __construct($type = '') {
```
typehint

rullzer reviewed (commented):
I don't see a reason for much of this implementation to be in the server. Like the actual implementation of a lot of those classes can just be in the dashboard app itself right?

ChristophWurst commented:
General note: you could and probably should remove most of the setters. How values are set in the implementation is not a concern of the server or other components, these are internals. All that matters to the server is how to get values (css, script, HTML, function references), right? That would shrink the interface sizes *a lot*.

ChristophWurst commented:
- [ ] Unclear: How do `IWidgetSetting` and `WidgetSettings` fit together? Why is there an implementation for it, if there's no API to get an instance of the interface? This only makes sense if there is some kind of factory method to obtain an instance. Same goes for `WidgetSetup`
- [ ] Unclear: Are `IWidgetSetup`, `IWidgetSetting` and `IWidgetTemplate` meant to be implemented by other apps? Do we rely on them being serializable? Then the `JsonSerializable` should be an interface those interfaces extend.

ChristophWurst commented on lib/public/Dashboard/Service/IEventsService.php:
```diff
@@ -0,0 +1,78 @@
+	 * @param array $users
+	 * @param array $payload
+	 * @param string $uniqueId
+	 * @return mixed
```
`IWidgetEvent`?

ChristophWurst commented on lib/private/Dashboard/DashboardManager.php:
```diff
@@ -0,0 +1,168 @@
+	private $appManager;
+
+	/** @var IServerContainer */
+	private $container;
```
unused

ChristophWurst commented on lib/private/Dashboard/DashboardManager.php:
```diff
@@ -0,0 +1,168 @@
+
+
+	/** @var IAppManager */
+	private $appManager;
```
unused

ChristophWurst commented on lib/private/Dashboard/Model/WidgetSetting.php:
```diff
@@ -0,0 +1,238 @@
+ *
+ * @package OC\Dashboard\Model
+ */
+class WidgetSetting implements JsonSerializable {
```
I think this implementation can be safely move to the dashboard app

ChristophWurst commented on lib/private/Dashboard/Model/WidgetSetup.php:
```diff
@@ -0,0 +1,269 @@
+ *
+ * @package OC\Dashboard\Model
+ */
+class WidgetSetup implements JsonSerializable {
```
I think this implementation can be safely move to the dashboard app

ChristophWurst commented on lib/private/Dashboard/Model/WidgetTemplate.php:
```diff
@@ -0,0 +1,317 @@
+ *
+ * @package OC\Dashboard\Model
+ */
+class WidgetTemplate implements JsonSerializable {
```
I think this implementation can be safely move to the dashboard app

ChristophWurst reviewed (changes requested):
Looks a lot better :+1: 

Let's move the implementations to the dashboard app. I don't think they should be in the server as they aren't even used directly.

Nitpick: `@see WidgetSetting` is a bit misleading. Not sure what it should tell me. It's also a bit misplaced in the interfaces because that would point to the private implementation. All usage should be documented in the public interfaces though :wink:

daita commented:
@ChristophWurst 

- I removed the unused variable from DashboardManager

- WidgetSetting, WidgetSetup and WidgetTemplate are created by the app that generate the widget, so it should stays here. This should be documented in the IDashboardWidget

ChristophWurst commented:
> WidgetSetting, WidgetSetup and WidgetTemplate are created by the app that generate the widget, so it should stays here. This should be documented in the IDashboardWidget

WidgetSettings is not used outside the private namespace. Hence it must not be used by apps.

`IDashboardWidget::getWidgetSetup` returns a `WidgetSetup`, which violates the public/private API separation. It should instead return an `IWidgetSetup`. Fixing this will effectively make the `WidgetSetup` implementation independent of server and apps, hence moving it to the dashboard app is fine/recommended.

ChristophWurst commented:
>     * WidgetSetting, WidgetSetup and WidgetTemplate are created by the app that generate the widget, so it should stays here. This should be documented in the IDashboardWidget

If a construct (class/interface) is used by an app, it has to be a public API. There are two possible use cases:

* The server only cares about an objects methods: create an interface that is implemented in an app's class.
* The app should extend an existing (potentially abstract) class: export the class in the public API
* The app use objects generated by the server or dashboard app: extract an interface from those interfaces, export them in the public API and provide factory classes/methods to create the objects as public API.

ChristophWurst commented:
After a bit of discussion I think I've finally understood how the different components will be used. I created a little diagram that outlines where certain classes/interfaces live:
![bildschirmfoto von 2018-10-12 13-08-30](https://user-images.githubusercontent.com/1374172/46866121-5f6c2b80-ce20-11e8-919f-d283b52fae03.png)

This works for @daita and clearly separates what should be in server/dashboard and other apps. This removes and dependency from server on the dashboard app. The implementation part in server (esp the private part) is minimal. The three classes for setup, settings and template were deliberately moved to OCP as classes as they are simple models (or better *data transfer objects*).

rullzer commented:
> After a bit of discussion I think I've finally understood how the different components will be used. I created a little diagram that outlines where certain classes/interfaces live:
> ![bildschirmfoto von 2018-10-12 13-08-30](https://user-images.githubusercontent.com/1374172/46866121-5f6c2b80-ce20-11e8-919f-d283b52fae03.png)
> 
> This works for @daita and clearly separates what should be in server/dashboard and other apps. This removes and dependency from server on the dashboard app. The implementation part in server (esp the private part) is minimal. The three classes for setup, settings and template were deliberately moved to OCP as classes as they are simple models (or better _data transfer objects_).

Yes makes sense.
As some of the classes are only tiny and there to transport data. I would like to make the getters and setters at least final. We can always losen this later. But lets start from a strict implementation.

rullzer reviewed (approved):
Ok I think it is time to get this in.
It looks sane to me.

Lets see how far we get.
@daita could you squash your commits?

ChristophWurst reviewed (changes requested):
Wasn't the plan to have `WidgetSettings`, `WidgetSetup` and `WidgetTemplate` in OCP? They're in the private namespace right now.

ChristophWurst reviewed (approved):
Looks good now!

You probably have to run `./build/autoloaderchecker.sh` to make CI happy.

Also, please make squash this and make this a single commit change :wink:

MorrisJobke commented:
CI says no: https://drone.nextcloud.com/nextcloud/server/11524/93

ChristophWurst commented on lib/composer/composer/autoload_classmap.php:
```diff
@@ -122,6 +122,16 @@
+    'OCP\\Dashboard\\IDashboardManager' => $baseDir . '/lib/public/Dashboard/IDashboardManager.php',
+    'OCP\\Dashboard\\IDashboardWidget' => $baseDir . '/lib/public/Dashboard/IDashboardWidget.php',
+    'OCP\\Dashboard\\Model\\IWidgetConfig' => $baseDir . '/lib/public/Dashboard/Model/IWidgetConfig.php',
+    'OCP\\Dashboard\\Model\\IWidgetRequest' => $baseDir . '/lib/public/Dashboard/Model/IWidgetRequest.php',
```
interfaces were removed

ChristophWurst reviewed (changes requested):
Please remove additional files and rerun autoloader command

ChristophWurst reviewed (approved):
Nvm the last review, I was confused by the interface names it seems

ArtificialOwl committed:
include dashboard into core

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
Merge remote-tracking branch 'origin/interface-dashboard' into interface-dashboard

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
moving data object to OCP

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
update autoload files

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
+@since

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
